File: internal/content/generator.go
package content

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/reek004/MarloweQuill/internal/api"
)

type ContentGenerator struct {
	geminiClient *api.GeminiClient
}

type ContentType string

const (
	BlogPost     ContentType = "blog_post"
	Article      ContentType = "article"
	SocialMedia  ContentType = "social_media"
	Script       ContentType = "script"
)

type GenerationOptions struct {
	ContentType    ContentType
	Topic          string
	Tone           string
	Length         int
	AdditionalContext string
}

func NewContentGenerator(client *api.GeminiClient) *ContentGenerator {
	return &ContentGenerator{
		geminiClient: client,
	}
}

func (g *ContentGenerator) GenerateContent(opts GenerationOptions) (string, error) {
	// Construct prompt based on content type and options
	prompt := g.constructPrompt(opts)

	// Generate content using Gemini API
	content, err := g.geminiClient.GenerateContent(prompt)
	if err != nil {
		return "", err
	}

	return content, nil
}

func (g *ContentGenerator) constructPrompt(opts GenerationOptions) string {
	basePrompt := fmt.Sprintf("Write a %s about '%s'", opts.ContentType, opts.Topic)
	
	if opts.Tone != "" {
		basePrompt += fmt.Sprintf(" in a %s tone", opts.Tone)
	}
	
	if opts.Length > 0 {
		basePrompt += fmt.Sprintf(". Aim for approximately %d words", opts.Length)
	}
	
	if opts.AdditionalContext != "" {
		basePrompt += fmt.Sprintf(". Additional context: %s", opts.AdditionalContext)
	}
	
	return basePrompt
}

func (g *ContentGenerator) SaveContent(content string, contentType ContentType) (string, error) {
	// Create output directory if it doesn't exist
	outputDir := "generated_content"
	err := os.MkdirAll(outputDir, os.ModePerm)
	if err != nil {
		return "", err
	}

	// Generate filename with timestamp
	timestamp := time.Now().Format("20060102_150405")
	filename := filepath.Join(outputDir, fmt.Sprintf("%s_%s.txt", contentType, timestamp))

	// Write content to file
	err = os.WriteFile(filename, []byte(content), 0644)
	if err != nil {
		return "", err
	}

	return filename, nil
}

File: src/content/generator.ts
import fs from 'fs/promises';
import path from 'path';
import { GeminiClient } from '../api/gemini';

export type ContentType = 'blog_post' | 'article' | 'social_media' | 'script';

export interface GenerationOptions {
  contentType: ContentType | string;
  topic: string;
  tone?: string;
  length?: number;
  additionalContext?: string;
}

export class ContentGenerator {
  constructor(private geminiClient: GeminiClient) {}

  async generateContent(opts: GenerationOptions): Promise<string> {
    const prompt = this.constructPrompt(opts);
    return await this.geminiClient.generateContent(prompt);
  }

  private constructPrompt(opts: GenerationOptions): string {
    let prompt = `Write a ${opts.contentType} about '${opts.topic}'`;
    
    if (opts.tone) {
      prompt += ` in a ${opts.tone} tone`;
    }
    
    if (opts.length) {
      prompt += `. Aim for approximately ${opts.length} words`;
    }
    
    if (opts.additionalContext) {
      prompt += `. Additional context: ${opts.additionalContext}`;
    }
    
    return prompt;
  }

  async saveContent(content: string, contentType: string): Promise<string> {
    const outputDir = 'generated_content';
    await fs.mkdir(outputDir, { recursive: true });

    const timestamp = new Date().toISOString().replace(/[:\.]/g, '').slice(0, 15);
    const filename = path.join(outputDir, `${contentType}_${timestamp}.txt`);

    await fs.writeFile(filename, content, 'utf-8');
    return filename;
  }
}